File: Assets/MakeHuman/Icons/KsSoft/Plugins/Protocol/t_Billboard.cs
//==============================================================================================
/*!
	@file  KsSoft Protocol
	
	(counter SJIS string 京)
*/
//==============================================================================================
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

using KS;
//==========================================================================
/*!
	@brief	t_Billboard
*/
public class t_Billboard : ISerializable {
	//==========================================================================
	/*!
		@brief	Member
	*/
	public uint	m_mId;
	public uint	m_mMaterial;
	public uint	m_mLink;
	public uint	m_fcBone;
	public uint	m_origin;
	public uint	m_flag;
	public float	m_yoffset;
	public float	m_width;
	public float	m_height;
	public float	m_rotate;
	public t_BillboardFrame[]	aFrame;
	//==========================================================================
	/*!
		@brief	Constructor
	*/
	public t_Billboard() {
		clear();
	}
	//==========================================================================
	/*!
		@brief	Accessor
	*/
	public void clear() {
		m_mId = 0;
		m_mMaterial = 0;
		m_mLink = 0;
		m_fcBone = 0;
		m_origin = 0;
		m_flag = 0;
		m_yoffset = 0.0f;
		m_width = 0.0f;
		m_height = 0.0f;
		m_rotate = 0.0f;
		aFrame = new t_BillboardFrame[0];
	}
	public bool read(CReadVariable cVariable) {
		try {
			m_mId = cVariable.getU32();
			m_mMaterial = cVariable.getU32();
			m_mLink = cVariable.getU32();
			m_fcBone = cVariable.getU32();
			m_origin = cVariable.getU32();
			m_flag = cVariable.getU32();
			m_yoffset = cVariable.getFloat();
			m_width = cVariable.getFloat();
			m_height = cVariable.getFloat();
			m_rotate = cVariable.getFloat();
			{
				int n = (int) cVariable.getU8();
				if (n > 255) {
					cVariable.error("array size error in aFrame");
					return false;
				}
				aFrame = new t_BillboardFrame [n];
				for(int i = 0;i < n;++i) {
					aFrame[i] = new t_BillboardFrame();
					if (!aFrame[i].read(cVariable)) return false;
				}
			}
		} catch (System.Exception e) {
			Debug.LogError(e);return false;
		}
		return true;
	}
	public bool write(CWriteVariable cVariable) {
		cVariable.put(m_mId);
		cVariable.put(m_mMaterial);
		cVariable.put(m_mLink);
		cVariable.put(m_fcBone);
		cVariable.put(m_origin);
		cVariable.put(m_flag);
		cVariable.put(m_yoffset);
		cVariable.put(m_width);
		cVariable.put(m_height);
		cVariable.put(m_rotate);
		{	
				if (aFrame.Length > 255) {
					cVariable.error("array size error in aFrame");
					return false;
				}
			int n = aFrame.Length;
			cVariable.put((byte) n);
			for(int i = 0;i < n;++i) {
				if (!aFrame[i].write(cVariable)) return false;

			}
		}
		return true;
	}
};

File: Assets/MakeHuman/Icons/KsSoft/Editor/Multilingual/tools/protocol/t_Billboard.py
from CVariable import CReadVariable,CWriteVariable
from Vector import *
from t_BillboardFrame import	t_BillboardFrame
#/==========================================================================
#*!
#	@brief	t_Billboard
#/
class t_Billboard:
	#/==========================================================================
	#*!
	#	@brief	Member
	#/
	m_mId = 0	#U32
	m_mMaterial = 0	#U32
	m_mLink = 0	#U32
	m_fcBone = 0	#U32
	m_origin = 0	#U32
	m_flag = 0	#U32
	m_yoffset = 0.0	#float
	m_width = 0.0	#float
	m_height = 0.0	#float
	m_rotate = 0.0	#float
	aFrame = []	#t_BillboardFrame
	#/==========================================================================
	#*!
	#	@brief	Constructor
	#/
	def __init__(self):
		self.clear()

	#/==========================================================================
	#*!
	#	@brief	Accessor
	#/
	def clear(self):
		self.m_mId = 0
		self.m_mMaterial = 0
		self.m_mLink = 0
		self.m_fcBone = 0
		self.m_origin = 0
		self.m_flag = 0
		self.m_yoffset = 0.0
		self.m_width = 0.0
		self.m_height = 0.0
		self.m_rotate = 0.0
		self.aFrame = []

	def read(self,cVariable):
		self.m_mId = cVariable.getU32()
		self.m_mMaterial = cVariable.getU32()
		self.m_mLink = cVariable.getU32()
		self.m_fcBone = cVariable.getU32()
		self.m_origin = cVariable.getU32()
		self.m_flag = cVariable.getU32()
		self.m_yoffset = cVariable.getFloat()
		self.m_width = cVariable.getFloat()
		self.m_height = cVariable.getFloat()
		self.m_rotate = cVariable.getFloat()
		n = cVariable.getU8()
		if (n > 255) :
			cVariable.error('array size error in aFrame')
			return False
		self.aFrame = [None]*n
		for i in range(n):
			self.aFrame[i] = t_BillboardFrame()
			if (not self.aFrame[i].read(cVariable)):
				return False
		return True

	def write(self,cVariable):
		cVariable.putU32(self.m_mId)
		cVariable.putU32(self.m_mMaterial)
		cVariable.putU32(self.m_mLink)
		cVariable.putU32(self.m_fcBone)
		cVariable.putU32(self.m_origin)
		cVariable.putU32(self.m_flag)
		cVariable.putFloat(self.m_yoffset)
		cVariable.putFloat(self.m_width)
		cVariable.putFloat(self.m_height)
		cVariable.putFloat(self.m_rotate)
		if (len(self.aFrame) > 255) :
			cVariable.error('array size error in aFrame')
			return False
		n = len(self.aFrame)
		cVariable.putU8(n)
		for i in range(n):
			if (not self.aFrame[i].write(cVariable)):
				return False
		return True

